File: src/main/java/at/ac/uibk/dps/cirrina/classes/transition/TransitionClassBuilder.java
package at.ac.uibk.dps.cirrina.classes.transition;

import at.ac.uibk.dps.cirrina.classes.statemachine.StateMachineClassBuilder;
import at.ac.uibk.dps.cirrina.csm.Csml.ActionDescription;
import at.ac.uibk.dps.cirrina.csm.Csml.TransitionDescription;
import at.ac.uibk.dps.cirrina.execution.object.action.Action;
import at.ac.uibk.dps.cirrina.execution.object.action.ActionBuilder;
import at.ac.uibk.dps.cirrina.execution.object.guard.Guard;
import at.ac.uibk.dps.cirrina.execution.object.guard.GuardBuilder;
import java.util.List;
import java.util.function.Function;

/**
 * Abstract transitionClass builder.
 */
public abstract class TransitionClassBuilder {

  /**
   * Construct a builder from a transition description.
   *
   * @param transitionDescription Transition description.
   * @return Builder.
   */
  public static TransitionClassBuilder from(TransitionDescription transitionDescription) {
    return new TransitionClassFromDescriptionBuilder(transitionDescription);
  }

  /**
   * Builds the transitionClass.
   *
   * @return the transitionClass.
   * @throws IllegalArgumentException In case the transitionClass could not be built.
   */
  public abstract TransitionClass build() throws IllegalArgumentException;

  /**
   * TransitionClass builder implementation. Builds a transitionClass based on a transitionClass class.
   *
   * @see StateMachineClassBuilder
   */
  private static final class TransitionClassFromDescriptionBuilder extends TransitionClassBuilder {

    /**
     * Transition description.
     */
    private final TransitionDescription transitionDescription;

    /**
     * Initializes this builder.
     *
     * @param transitionDescription Transition description.
     */
    private TransitionClassFromDescriptionBuilder(TransitionDescription transitionDescription) {
      this.transitionDescription = transitionDescription;
    }

    /**
     * Builds the transition class.
     *
     * @return Transition class.
     * @throws IllegalArgumentException If a guard name does not exist.
     * @throws IllegalArgumentException If an action name does not exist.
     */
    @Override
    public TransitionClass build() throws IllegalArgumentException {
      // Resolve guards
      Function<List<String>, List<Guard>> resolveGuards = (List<String> guards) ->
        guards
          .stream()
          .map(expression -> GuardBuilder.from(expression).build())
          .toList();

      // Resolve actions
      Function<List<? extends ActionDescription>, List<Action>> resolveActions = (List<
        ? extends ActionDescription
      > actions) ->
        actions
          .stream()
          .map(actionClass -> ActionBuilder.from(actionClass).build())
          .toList();

      // Create the appropriate transitionClass
      // KLUDGE: The CSML language used to have both on-transitions and transitions in the language
      // itself, we simplified this, but the OnTransitionClass and TransitionClass remain to make
      // this a minimal change at the time of writing
      // TODO: Unify OnTransitionClass and TransitionClass
      if (transitionDescription.getEvent() != null) {
        return new OnTransitionClass(
          transitionDescription.getTarget(),
          transitionDescription.getOr(),
          resolveGuards.apply(transitionDescription.getGuards()),
          resolveActions.apply(transitionDescription.getActions()),
          transitionDescription.getEvent()
        );
      } else {
        return new TransitionClass(
          transitionDescription.getTarget(),
          transitionDescription.getOr(),
          resolveGuards.apply(transitionDescription.getGuards()),
          resolveActions.apply(transitionDescription.getActions())
        );
      }
    }
  }
}

File: src/main/java/at/ac/uibk/dps/cirrina/classes/transition/TransitionClassBuilder.kt
package at.ac.uibk.dps.cirrina.classes.transition

import at.ac.uibk.dps.cirrina.csm.Csml.TransitionDescription
import at.ac.uibk.dps.cirrina.execution.`object`.action.ActionBuilder
import at.ac.uibk.dps.cirrina.execution.`object`.guard.GuardBuilder

/**
 * A functional builder for constructing [TransitionClass] instances from [TransitionDescription]s.
 *
 * @property transitionDescription the source for the transition.
 * @property event the optional event name that triggers this transition.
 */
class TransitionClassBuilder
private constructor(
  private val transitionDescription: TransitionDescription,
  private val event: String? = null,
) {

  /**
   * Associates a specific [event] name with this transition and returns a new builder instance.
   *
   * @param event the identifier of the event to trigger this transition.
   * @return a new builder instance containing the specified event.
   */
  fun withEvent(event: String): TransitionClassBuilder =
    TransitionClassBuilder(transitionDescription, event)

  /**
   * Builds a [TransitionClass].
   *
   * @return a [Result] containing the fully constructed [TransitionClass].
   */
  fun build(): Result<TransitionClass> = runCatching {
    TransitionClass(
      event = event,
      to = transitionDescription.to,
      iif = transitionDescription.iif?.let { GuardBuilder.from(it).build().getOrThrow() },
      `do` = transitionDescription.`do`.map { ActionBuilder.from(it).build().getOrThrow() },
      or = transitionDescription.or,
    )
  }

  companion object {
    /**
     * Creates a starting [TransitionClassBuilder] from the provided [transitionDescription].
     *
     * @param transitionDescription the source description.
     * @return a new builder instance.
     */
    fun from(transitionDescription: TransitionDescription): TransitionClassBuilder =
      TransitionClassBuilder(transitionDescription)
  }
}
